src/util/dsp/window.cpp

```diff
@@ -0,0 +1,125 @@
+#include "window.hpp"
+#include <math.h>
+
+namespace otto::util::dsp {
+
+  Window::Window (int size, WindowType type, bool normalize) : windowBuffer(size)
+  {
+      reset (size, type, normalize);
+  }
+
+
+  void Window::reset (int size, WindowType type, bool normalize) noexcept
+  {
+      windowBuffer.resize (size);
+      compute (windowBuffer.begin(), size, type, normalize);
+  }
+
+
+  void Window::compute (double* buffer, int size, WindowType type, bool normalize) noexcept
+  {
+      auto cosn = [](int n, int i, int size) { return cos (static_cast<double> (n * M_PI * i) / static_cast<double> (size - 1)); };
+
+      switch (type)
+      {
+          case rectangular:
+          {
+              for (int i = 0; i < size; ++i)
+                  buffer[i] = 1.0;
+          }
+          break;
+
+          case triangular:
+          {
+              auto middlePosition = 0.5 * static_cast<double> (size - 1);
+
+              for (int i = 0; i < size; ++i)
+                  buffer[i] = 1.0 - std::abs ((static_cast<double> (i) - middlePosition) / middlePosition);
+          }
+          break;
+
+          case hann:
+          {
+              for (int i = 0; i < size; ++i)
+              {
+                  auto cos2 = cosn (2, i, size);
```

Comment:
`std::cosn`

Reply:
cosn is defined earlier, it's not a standard function.